components/scream/src/physics/p3/tests/p3_unit_tests.cpp

```diff
@@ -451,6 +451,186 @@ static void  cloud_water_autoconversion_unit_bfb_tests(){
 
   };//TestP3UpdatePrognosticIce
 
+  template <typename D>
+  struct UnitWrap::UnitTest<D>::TestP3UpdatePrognosticLiq
+  {
+    static void  update_prognostic_liquid_unit_bfb_tests(){
+
+      static constexpr Int max_pack_size = 16;
+
+      REQUIRE(Spack::n <= max_pack_size);
+
+      //fortran generated data is input to the following
+      P3UpdatePrognosticLiqData pupldc[max_pack_size] = {
+
+	{1.0631E-12, 1.0631E+00, 1.5833E-12, 1.5833E+00, 0.0000E+00, 2.4190E-02, 0.0000E+00, 0.0000E+00, 0.0000E+00, 4.2517E+00,
+	 true      , 8.6718E-01, 1.0037E+00, 2.5010E+06, 1.8000E+03, 2.9902E+02, 5.0000E-02, 1.0000E-06, 1.0000E+06, 1.0010E-06,
+	 6.3726E+05},
+
+	{3.2784E-08, 1.8780E+07, 2.1753E-11, 1.2461E+04, 0.0000E+00, 7.8657E+03, 0.0000E+00, 0.0000E+00, 0.0000E+00, 5.8748E+04,
+	 true      , 9.8387E-01, 1.0741E+00, 2.5010E+06, 1.8000E+03, 2.9033E+02, 3.7211E-03, 5.9050E-05,-6.6723E+09,-5.9050E-05,
+	 -8.6159E+07},
+
+	{3.2796E-09, 1.8778E+07, 1.8830E-12, 1.0782E+04, 0.0000E+00, 6.8061E+03, 0.0000E+00, 0.0000E+00, 0.0000E+00, 6.3698E+04,
```

Comment:
It looks like log_predictNC=false was never tested. In general, we want to tweak our input data to exercise as many branch paths as possible.

Reply:
Thanks for this suggestion. I tried alternating `true` and `false` for each pack in `pupldc` data structure. It looks like, the unit test only picks value for this bool mentioned in the first pack of `pupldc` (lines 466 and 467 above - which is `true` in this case). This is because I have the following at line 555:
`log_predictNc = pupldc_device(0).log_predictNc;`

For the unit test to pick value from each pack (rather than just the first pack), do I need to declare this logical as `Smask` rather than `bool`?

Reply:
@singhbalwinder great point, it's easy to forget that Cxx does all this input with a single call. I withdraw my original comment.

Reply:
Thanks. Toggling this to false and true forced me to check if the results agree between Fortran and C (and they do) for these conditions. So it was a good exercise.